plugins/inputs/elasticsearch/elasticsearch.go

```diff
@@ -186,25 +195,44 @@ func (e *Elasticsearch) Gather(acc telegraf.Accumulator) error {
 		e.client = client
 	}
 
-	var wg sync.WaitGroup
-	wg.Add(len(e.Servers))
+	if e.ClusterStats {
+		e.serverInfo = make(map[string]serverInfo)
+		for _, serv := range e.Servers {
+			go func(s string, acc telegraf.Accumulator) {
+				info := serverInfo{}
 
-	for _, serv := range e.Servers {
-		go func(s string, acc telegraf.Accumulator) {
-			defer wg.Done()
-			url := e.nodeStatsUrl(s)
-			e.isMaster = false
+				var err error
+
+				// Gather node ID
+				if info.nodeID, err = e.gatherNodeID(s + "/_nodes/_local/name"); err != nil {
+					acc.AddError(fmt.Errorf(mask.ReplaceAllString(err.Error(), "http(s)://XXX:XXX@")))
+					return
+				}
 
-			if e.ClusterStats {
 				// get cat/master information here so NodeStats can determine
 				// whether this node is the Master
-				if err := e.setCatMaster(s + "/_cat/master"); err != nil {
+				if info.masterID, err = e.getCatMaster(s + "/_cat/master"); err != nil {
 					acc.AddError(fmt.Errorf(mask.ReplaceAllString(err.Error(), "http(s)://XXX:XXX@")))
 					return
 				}
-			}
 
-			// Always gather node states
+				e.serverInfoMutex.Lock()
+				e.serverInfo[s] = info
+				e.serverInfoMutex.Unlock()
```

Comment:
I notice now that we don't wait for these goroutines to complete before moving on to the next section of the code, this means that we still might be writing the serverInfo by the time we try to read from it later on.  It is possible that we still haven't figured out who the master is on line 252.

It looks like we will need to use a sync.WaitGroup and then wait for all goroutines in this loop before moving into the next for loop.

Reply:
Correct! I missed that indeed. Fixed now.